Charting updates and fixes

mbest opened the pull request:
#### Pull request checklist

- [x] Include a change request file using `$ yarn change`

#### Description of changes

- When `hideLegend` is true, no space is reserved for the legend
- turn on `strictBindCallApply` compiler option and fix resulting errors
- For VerticalStackedBarChart, support turning off hover action on legends
- VerticalStackedBarChart: add minimum bar height option
- VerticalStackedBarChart: add `onBarClick` callback option
- VerticalStackedBarChart: simplify how data is passed to callbacks
- VerticalStackedBarChart: don't output empty stack `<g>`
- VerticalStackedBarChart: disallow `yMinValue` since setting it breaks the scale of the chart

shivasai09 commented on packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:
```diff
@@ -213,25 +210,23 @@ export class VerticalStackedBarChartBase extends React.Component<
-          if (line.index === undefined) {
-            line.index = index;
-          }
+        item.lineData.forEach(line => {
```
could you please revert this change. we need to set `x` value and `index` value only when it is undefined, this new code will override it. which we don't want

mbest replied:
My new code makes a copy of the objects, so it will need to set these each time this function is run.

mbest replied:
BTW, the previous code was modifying the objects (props) passed into this component, which is generally frowned upon.

Raghurk replied:
@shivasai09 can you make sure this does not break our code?

shivasai09 replied:
sure.

shivasai09 replied:
i have noticed one issue. and mentioned in below comments

shivasai09 commented on packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:
```diff
@@ -253,8 +248,12 @@ export class VerticalStackedBarChartBase extends React.Component<
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  private _createLines = (xScale: any, yScale: any, containerHeight: number, containerWidth: number): JSX.Element => {
+  private _createLines = (
+    xScale: NumericScale,
```
it can also be dateScale

mbest replied:
For strings, it still uses a numeric scale using the indexes. It cannot be dateScale

shivasai09 commented on packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:
```diff
@@ -299,27 +298,26 @@ export class VerticalStackedBarChartBase extends React.Component<
-                      ?.lineData,
-                  )
+                ? this._lineHover.bind(this, circlePoint)
+                : this._onStackHover.bind(this, circlePoint.x)
```
`circlePoint.x`?? you need to send the StackData right? this will break the hover on circle i belive

shivasai09 replied:
![image](https://user-images.githubusercontent.com/33802398/98070970-4f509a00-1e88-11eb-9672-dcdbc23ca589.png)

mbest replied:
Nope. It works fine
![image](https://user-images.githubusercontent.com/1230364/98294583-091d3880-1f65-11eb-8eea-87aed75a1140.png)

shivasai09 replied:
no it isn't
![image](https://user-images.githubusercontent.com/33802398/98312155-1c78e400-1ff7-11eb-99c2-1ff6a0782c4c.png)

shivasai09 commented on packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:
```diff
@@ -213,25 +210,23 @@ export class VerticalStackedBarChartBase extends React.Component<
+          linesData.push({
+            ...line,
+            index,
+            x: item,
```
@mbest  why are you setting `item` to `x` instead of `x.xAxisPoint`

mbest replied:
Using `xAxisPoint` means having to do a `find` later. This change directly links to the item. I've renamed the property to make it a bit more informative.

shivasai09 commented on packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:
```diff
@@ -299,27 +298,26 @@ export class VerticalStackedBarChartBase extends React.Component<
-                      ?.lineData,
-                  )
+                ? this._lineHover.bind(this, circlePoint)
+                : this._onStackHover.bind(this, circlePoint.x)
```
@mbest 

it's not working fine
![image](https://user-images.githubusercontent.com/33802398/98312724-539bc500-1ff8-11eb-8a7e-c3c83fda2277.png)

mbest replied:
Fixed it: 
![image](https://user-images.githubusercontent.com/1230364/98315229-9d4ec600-1f8c-11eb-865d-d90fa37d5e13.png)

shivasai09 commented:
@mbest , `_onStackHover` and `_onStackFocus` looks almost similar, can you make it reusable. since this pr is doing clean-up, it would be appropriate to make this change..

what say?

mbest commented:
> @mbest , `_onStackHover` and `_onStackFocus` looks almost similar, can you make it reusable. since this pr is doing clean-up, it would be appropriate to make this change..

Sure. I also did the same for the "rect" versions.

mbest commented:
> @mbest , `_onStackHover` and `_onStackFocus` looks almost similar, can you make it reusable. since this pr is doing clean-up, it would be appropriate to make this change..

Sure. I also did the same for the "rect" versions.

shivasai09 commented on packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:
```diff
@@ -559,82 +565,46 @@ export class VerticalStackedBarChartBase extends React.Component<
-      hoverXValue: xAxisPoint,
-      stackCalloutProps: found!,
-      activeXAxisDataPoint: xAxisPoint,
+        ? [...lineData!, ...stack.chartData.slice().reverse()]
```
hey @mbest , i think while taking master to your branch , you have reverted the behavior which we merged in https://github.com/microsoft/fluentui/pull/15814/files.. we were sorting the `lineData` in descending order

could you please put it back.?

Thanks,
For making the code clean up on this.

mbest replied:
Merged now.

shivasai09 replied:
hey, can you see line data sorted.?

i am not able to [see](http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/15816/merge/charting/dist/index.html#/examples/VerticalStackedBarChart), 
![image](https://user-images.githubusercontent.com/33802398/98621931-67179a80-232e-11eb-959d-37d2efc7f5e3.png)


it should be in descending. like yellow should be on last and pink should be on first


if it's working for you in local, can you post a pic here. then i guess we are good merge this pr

mbest replied:
![image](https://user-images.githubusercontent.com/1230364/98657049-4b909d80-22f6-11eb-9c32-926c1e7b3200.png)

shivasai09 commented on packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:
```diff
@@ -559,82 +565,46 @@ export class VerticalStackedBarChartBase extends React.Component<
-        item.shouldDrawBorderBottom = true;
-      });
+  private _onStackFocus(stack: IVerticalStackedChartProps, groupRef: IRefArrayData): void {
+    if (groupRef.refElement) {
```
is this if condition really necessary ?

shivasai09 commented:
@mbest  once i hover on any legend, and then if i mouse out  on that , it stays in hover state only
![image](https://user-images.githubusercontent.com/33802398/98658192-c98d8c80-2368-11eb-9ebd-ab9279ac346f.png)

This happens only in legends. but for line chart and all it works fine.
why are you changing this behavior?

mbest commented:
> @mbest once i hover on any legend, and then if i mouse out on that , it stays in hover state only
> This happens only in legends. but for line chart and all it works fine.
> why are you changing this behavior?

The actual effect of legend interaction happens due to the callback functions provided. I changed `Legend` to update the UX only when there's also a callback for that interaction, because I needed to add support for not reacting to hover or focus.

To ensure the `Legend` examples work as they did, I've added `onMouseOutAction` callbacks to them.
